standardize docs and minor function cleanup

jkvyff opened the pull request:
Ref: #222 
standardize terms, return non-returning reduce, remove unused args

evanplaice commented on src/arrays/pull.js:
```diff
@@ -2,7 +2,7 @@
  *
  * @param {Array} array input array
- * @param {...*} values values to be removed from the array
+ * @param {...*} [values] values to be removed from the array
```
Rest params can't be optional. Typescript will reject this.

evanplaice commented on src/objects/assign.js:
```diff
@@ -3,7 +3,7 @@
  *
  * @param {object} object input object
- * @param {...object} sources input object/s
+ * @param {...object} [sources] input source object(s)
```
Rest params can't be optional

evanplaice commented on src/objects/at.js:
```diff
@@ -2,7 +2,7 @@
  *
  * @param {object} object input object
- * @param {...(string|string[])} paths strings describing paths to be returned from an object
+ * @param {...(string|string[])} [paths] strings describing paths to be returned from an object
```
Rest params can't be optional

evanplaice commented on src/objects/defaults.js:
```diff
@@ -3,7 +3,7 @@
  *
  * @param {object} object input object
- * @param {...object} sources input object/s
+ * @param {...object} [sources] input source object(s)
```
Rest params...

evanplaice commented on src/objects/defaultsDeep.js:
```diff
@@ -3,7 +3,7 @@
  *
  * @param {object} object input object
- * @param {...object} sources input object/s
+ * @param {...object} [sources] input source object(s)
```
Rest params...

jkvyff commented:
Getting errors on rebuild about "option 'allowJs' cannot be specified with option 'declaration". Does that need an extra flag?

evanplaice commented:
No, you need to update Typescript. IIRC, allowing both "allowJS" and "checkJS" was added in 3.6. This is what enabled the ability to generate typings directly from JSDoc.

#BleedingEdgeProblems

jkvyff commented:
Ah, I found it was mentioned a couple months ago. Should be good now.

evanplaice commented:
What's `identity` here?

```javascript
/*  * @param {function} [predicate = identity] optional function to test against object values */
```

Can we just leave the default value out altogether?

evanplaice commented:
IMO, `undefined` can be assumed to be the default. It doesn't need to appear in the docs.

```javascript
/*  * @param {*} [thisArg = undefined] optional this in a function call */
```

jkvyff commented:
Good catch, identity to me just meant returns each value of self when iterated, but that didn't become a practice for the library. In the places it could be used, I opted to just return the original object if nothing was happening to the object as it would be much faster. 

For the second one, thisArg being undefined is more of an implementation detail and probably wouldn't help a user understand the function any more.

Removed both from documentation, removed redundant set as undefined for thisArg in implementation.

evanplaice commented:
Is this ready to land?

jkvyff commented:
Yep yep
